File: dotnet/src/Envdoctor/Utils/Json.cs
using System.Globalization;
using System.Text;

namespace Envdoctor.Utils;

/// An insertion-ordered JSON object.
public sealed class JsonObject : List<KeyValuePair<string, object?>>
{
    public void Add(string key, object? value) => Add(new KeyValuePair<string, object?>(key, value));
}

/// A minimal JSON writer producing the exact byte shape shared by
/// `JSON.stringify(x, null, 2)` and `serde_json::to_string_pretty`: two-space
/// indentation, `": "` key separator, `{}`/`[]` for empty containers, and
/// control-character escaping with lowercase hex.
public static class Json
{
    public static string Pretty(object? value)
    {
        var sb = new StringBuilder();
        Write(sb, value, 0, pretty: true);
        return sb.ToString();
    }

    public static string Compact(object? value)
    {
        var sb = new StringBuilder();
        Write(sb, value, 0, pretty: false);
        return sb.ToString();
    }

    private static void Write(StringBuilder sb, object? value, int depth, bool pretty)
    {
        switch (value)
        {
            case null:
                sb.Append("null");
                break;
            case bool b:
                sb.Append(b ? "true" : "false");
                break;
            case string s:
                WriteString(sb, s);
                break;
            case int i:
                sb.Append(i.ToString(CultureInfo.InvariantCulture));
                break;
            case long l:
                sb.Append(l.ToString(CultureInfo.InvariantCulture));
                break;
            case double d:
                sb.Append(d.ToString("G17", CultureInfo.InvariantCulture));
                break;
            case JsonObject obj:
                WriteObject(sb, obj, depth, pretty);
                break;
            case IEnumerable<KeyValuePair<string, object?>> pairs:
                WriteObject(sb, pairs, depth, pretty);
                break;
            case System.Collections.IEnumerable list:
                WriteArray(sb, list.Cast<object?>(), depth, pretty);
                break;
            default:
                throw new ArgumentException($"Unsupported JSON value type: {value.GetType()}");
        }
    }

    private static void WriteObject(StringBuilder sb, IEnumerable<KeyValuePair<string, object?>> obj, int depth, bool pretty)
    {
        var entries = obj.ToList();
        if (entries.Count == 0)
        {
            sb.Append("{}");
            return;
        }
        sb.Append('{');
        for (var i = 0; i < entries.Count; i++)
        {
            if (pretty)
            {
                sb.Append('\n');
                Indent(sb, depth + 1);
            }
            WriteString(sb, entries[i].Key);
            sb.Append(pretty ? ": " : ":");
            Write(sb, entries[i].Value, depth + 1, pretty);
            if (i < entries.Count - 1)
                sb.Append(',');
        }
        if (pretty)
        {
            sb.Append('\n');
            Indent(sb, depth);
        }
        sb.Append('}');
    }

    private static void WriteArray(StringBuilder sb, IEnumerable<object?> items, int depth, bool pretty)
    {
        var list = items.ToList();
        if (list.Count == 0)
        {
            sb.Append("[]");
            return;
        }
        sb.Append('[');
        for (var i = 0; i < list.Count; i++)
        {
            if (pretty)
            {
                sb.Append('\n');
                Indent(sb, depth + 1);
            }
            Write(sb, list[i], depth + 1, pretty);
            if (i < list.Count - 1)
                sb.Append(',');
        }
        if (pretty)
        {
            sb.Append('\n');
            Indent(sb, depth);
        }
        sb.Append(']');
    }

    private static void Indent(StringBuilder sb, int depth)
    {
        for (var i = 0; i < depth; i++)
            sb.Append("  ");
    }

    private static void WriteString(StringBuilder sb, string s)
    {
        sb.Append('"');
        foreach (var c in s)
        {
            switch (c)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default:
                    if (c < ' ')
                        sb.Append("\\u").Append(((int)c).ToString("x4", CultureInfo.InvariantCulture));
                    else
                        sb.Append(c);
                    break;
            }
        }
        sb.Append('"');
    }
}

File: kotlin/src/main/kotlin/io/github/arun_skg/envdoctor/utils/Json.kt
package io.github.arun_skg.envdoctor.utils

/**
 * An insertion-ordered JSON object, matching the semantics of the reference's
 * ordered maps (ES2015 object insertion order / serde_json Map).
 */
class JsonObject : LinkedHashMap<String, Any?>()

/**
 * A minimal JSON writer producing the exact byte shape shared by
 * `JSON.stringify(x, null, 2)`: two-space indentation, `": "` key separator,
 * `{}`/`[]` for empty containers, and control-character escaping with
 * lowercase hex.
 */
object Json {
    fun pretty(value: Any?): String {
        val sb = StringBuilder()
        write(sb, value, 0, pretty = true)
        return sb.toString()
    }

    fun compact(value: Any?): String {
        val sb = StringBuilder()
        write(sb, value, 0, pretty = false)
        return sb.toString()
    }

    private fun write(sb: StringBuilder, value: Any?, depth: Int, pretty: Boolean) {
        when (value) {
            null -> sb.append("null")
            is Boolean -> sb.append(if (value) "true" else "false")
            is String -> writeString(sb, value)
            is Int -> sb.append(value.toString())
            is Long -> sb.append(value.toString())
            is Double -> sb.append(jsNumber(value))
            is JsonObject -> writeObject(sb, value.entries.map { it.key to it.value }, depth, pretty)
            is Map<*, *> ->
                writeObject(sb, value.entries.map { (k, v) -> (k.toString() to v) }, depth, pretty)
            is List<*> -> writeArray(sb, value, depth, pretty)
            is Array<*> -> writeArray(sb, value.asList(), depth, pretty)
            else -> throw IllegalArgumentException("Unsupported JSON value type: ${value.javaClass}")
        }
    }

    private fun writeObject(sb: StringBuilder, entries: List<Pair<String, Any?>>, depth: Int, pretty: Boolean) {
        if (entries.isEmpty()) {
            sb.append("{}")
            return
        }
        sb.append('{')
        for (i in entries.indices) {
            if (pretty) {
                sb.append('\n')
                indent(sb, depth + 1)
            }
            writeString(sb, entries[i].first)
            sb.append(if (pretty) ": " else ":")
            write(sb, entries[i].second, depth + 1, pretty)
            if (i < entries.size - 1) sb.append(',')
        }
        if (pretty) {
            sb.append('\n')
            indent(sb, depth)
        }
        sb.append('}')
    }

    private fun writeArray(sb: StringBuilder, items: List<*>, depth: Int, pretty: Boolean) {
        if (items.isEmpty()) {
            sb.append("[]")
            return
        }
        sb.append('[')
        for (i in items.indices) {
            if (pretty) {
                sb.append('\n')
                indent(sb, depth + 1)
            }
            write(sb, items[i], depth + 1, pretty)
            if (i < items.size - 1) sb.append(',')
        }
        if (pretty) {
            sb.append('\n')
            indent(sb, depth)
        }
        sb.append(']')
    }

    private fun indent(sb: StringBuilder, depth: Int) {
        for (i in 0 until depth) sb.append("  ")
    }

    private fun writeString(sb: StringBuilder, s: String) {
        sb.append('"')
        for (c in s) {
            when (c) {
                '"' -> sb.append("\\\"")
                '\\' -> sb.append("\\\\")
                '\b' -> sb.append("\\b")
                '\u000C' -> sb.append("\\f")
                '\n' -> sb.append("\\n")
                '\r' -> sb.append("\\r")
                '\t' -> sb.append("\\t")
                else ->
                    if (c < ' ') sb.append("\\u").append(c.code.toString(16).padStart(4, '0'))
                    else sb.append(c)
            }
        }
        sb.append('"')
    }

    /** JS number-to-string for the integral doubles YAML/JSON can produce. */
    fun jsNumber(d: Double): String {
        if (d == d.toLong().toDouble() && kotlin.math.abs(d) < 1e15) return d.toLong().toString()
        return d.toString()
    }
}
